# Análisis exploratorio inicial — Telco Customer Churn

Este notebook realiza la exploración inicial del dataset utilizado en el proyecto Sistema-MLOps.

El objetivo es identificar la estructura general de los datos, cantidad de registros y columnas, tipos de datos, valores nulos y distribución de la variable objetivo `Churn`.

In [3]:
import pandas as pd

df = pd.read_csv("../data/raw/churn.csv")

print("Dataset cargado correctamente")
print("Filas y columnas:", df.shape)

Dataset cargado correctamente
Filas y columnas: (7043, 21)


In [10]:
print("Número de filas:", df.shape[0])
print("Número de columnas:", df.shape[1])

Número de filas: 7043
Número de columnas: 21


In [11]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,...,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,...,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,...,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,...,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,...,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


In [12]:
print("Tipos de datos:")
print(df.dtypes)

Tipos de datos:
customerID              str
gender                  str
SeniorCitizen         int64
Partner                 str
Dependents              str
tenure                int64
PhoneService            str
MultipleLines           str
InternetService         str
OnlineSecurity          str
OnlineBackup            str
DeviceProtection        str
TechSupport             str
StreamingTV             str
StreamingMovies         str
Contract                str
PaperlessBilling        str
PaymentMethod           str
MonthlyCharges      float64
TotalCharges            str
Churn                   str
dtype: object


In [13]:
print("Valores nulos por columna:")
print(df.isnull().sum())

Valores nulos por columna:
customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


In [14]:
print("Distribución de Churn:")
print(df["Churn"].value_counts())

Distribución de Churn:
Churn
No     5174
Yes    1869
Name: count, dtype: int64


In [15]:
print(df.isnull().sum())

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64


In [ ]:
print("Distribución porcentual de Churn:")
print((df["Churn"].value_counts(normalize=True) * 100).round(2))

Distribución porcentual de Churn:
Churn
No     73.46
Yes    26.54
Name: proportion, dtype: float64


: 

## Conclusiones de la exploración inicial

- El dataset contiene 7043 registros y 21 columnas.
- La variable objetivo es `Churn`.
- La variable objetivo contiene las categorías `No` y `Yes`.
- La revisión inicial con Pandas no identificó valores nulos en las columnas.
- La distribución de la variable objetivo corresponde a 5174 registros con `Churn = No` y 1869 registros con `Churn = Yes`.
- Se identificaron los tipos de datos y la estructura general del dataset.
- Los resultados de esta exploración servirán como base para las siguientes etapas de limpieza y preparación de los datos.

## Variable objetivo

La variable objetivo oficial del proyecto es `Churn`.

Esta variable representa si un cliente abandona o no el servicio:

- `No`: el cliente no abandonó el servicio.
- `Yes`: el cliente abandonó el servicio.

Para las etapas de preparación, entrenamiento y evaluación del modelo, `Churn` será utilizada como variable objetivo (`target`).

In [4]:
variable_objetivo = "Churn"

print("Variable objetivo oficial:", variable_objetivo)
print("Valores encontrados:", df[variable_objetivo].unique().tolist())

Variable objetivo oficial: Churn
Valores encontrados: ['No', 'Yes']


# Limpieza de datos — S1-J4

En esta sección se realiza la limpieza inicial del dataset Telco Customer Churn.

Se revisarán:

- Registros duplicados.
- Columnas que no aportan información predictiva.
- Valores nulos.
- Valores vacíos o problemáticos.
- Tipos de datos inconsistentes.

El objetivo es obtener un conjunto de datos limpio y consistente para las siguientes etapas del modelado.

In [5]:
import pandas as pd

df = pd.read_csv("../data/raw/churn.csv")

print("Dimensión original:", df.shape)

Dimensión original: (7043, 21)


In [6]:
duplicados = df.duplicated().sum()

print("Registros duplicados:", duplicados)

Registros duplicados: 0


In [7]:
print("Valores únicos customerID:", df["customerID"].nunique())
print("Total de registros:", len(df))

Valores únicos customerID: 7043
Total de registros: 7043


In [8]:
print("Tipo de dato de TotalCharges:", df["TotalCharges"].dtype)

valores_vacios = df["TotalCharges"].astype(str).str.strip().eq("").sum()

print("Valores vacíos en TotalCharges:", valores_vacios)

Tipo de dato de TotalCharges: str
Valores vacíos en TotalCharges: 11


In [9]:
filas_totalcharges_vacio = df[
    df["TotalCharges"].astype(str).str.strip().eq("")
]

print("Cantidad de filas problemáticas:", len(filas_totalcharges_vacio))

filas_totalcharges_vacio[
    ["customerID", "tenure", "MonthlyCharges", "TotalCharges", "Churn"]
]

Cantidad de filas problemáticas: 11


,customerID,tenure,MonthlyCharges,TotalCharges,Churn
488,4472-LVYGI,0,52.55,,No
753,3115-CZMZD,0,20.25,,No
936,5709-LVOEQ,0,80.85,,No
1082,4367-NUYAO,0,25.75,,No
1340,1371-DWPAZ,0,56.05,,No
3331,7644-OMVMY,0,19.85,,No
3826,3213-VVOLG,0,25.35,,No
4380,2520-SGTTA,0,20.00,,No
5218,2923-ARZLG,0,19.70,,No
6670,4075-WKNIU,0,73.35,,No


## Tratamiento aplicado

Durante la revisión se identificaron 11 valores vacíos en la variable `TotalCharges`.

Los 11 registros presentan `tenure = 0`, lo que indica que corresponden a clientes sin antigüedad acumulada. Por esta razón, se decidió conservar los registros y reemplazar los valores vacíos de `TotalCharges` por `0`.

Posteriormente, `TotalCharges` se convierte a formato numérico.

También se elimina la columna `customerID` de la copia utilizada para modelado, debido a que contiene un identificador único para cada uno de los 7043 registros y no representa una característica predictiva del cliente.

No se encontraron registros duplicados.

In [10]:
df_limpio = df.copy()

print("Copia del dataset creada correctamente.")
print("Dimensión inicial:", df_limpio.shape)

Copia del dataset creada correctamente.
Dimensión inicial: (7043, 21)


In [12]:
df_limpio["TotalCharges"] = (
    df_limpio["TotalCharges"]
    .astype(str)
    .str.strip()
    .replace("", "0")
)

df_limpio["TotalCharges"] = pd.to_numeric(
    df_limpio["TotalCharges"],
    errors="coerce"
)

print("Tipo de TotalCharges:", df_limpio["TotalCharges"].dtype)
print("Nulos en TotalCharges:", df_limpio["TotalCharges"].isnull().sum())

Tipo de TotalCharges: float64
Nulos en TotalCharges: 0


In [13]:
df_limpio = df_limpio.drop(columns=["customerID"])

print("customerID existe:", "customerID" in df_limpio.columns)
print("Dimensión después de eliminar customerID:", df_limpio.shape)

customerID existe: False
Dimensión después de eliminar customerID: (7043, 20)


In [15]:
print("========== ANTES DE LA LIMPIEZA ==========")
print("Filas:", df.shape[0])
print("Columnas:", df.shape[1])
print("Duplicados reales en dataset original:", df.duplicated().sum())
print("customerID duplicados:", df["customerID"].duplicated().sum())
print("TotalCharges vacíos:", df["TotalCharges"].astype(str).str.strip().eq("").sum())
print("Tipo TotalCharges:", df["TotalCharges"].dtype)
print("customerID presente:", "customerID" in df.columns)

print("\n========== DESPUÉS DE LA LIMPIEZA ==========")
print("Filas:", df_limpio.shape[0])
print("Columnas:", df_limpio.shape[1])
print("Nulos totales:", df_limpio.isnull().sum().sum())
print("Nulos TotalCharges:", df_limpio["TotalCharges"].isnull().sum())
print("Tipo TotalCharges:", df_limpio["TotalCharges"].dtype)
print("customerID presente:", "customerID" in df_limpio.columns)

print(
    "Filas coincidentes después de eliminar customerID:",
    df_limpio.duplicated().sum()
)

========== ANTES DE LA LIMPIEZA ==========
Filas: 7043
Columnas: 21
Duplicados reales en dataset original: 0
customerID duplicados: 0
TotalCharges vacíos: 11
Tipo TotalCharges: str
customerID presente: True

========== DESPUÉS DE LA LIMPIEZA ==========
Filas: 7043
Columnas: 20
Nulos totales: 0
Nulos TotalCharges: 0
Tipo TotalCharges: float64
customerID presente: False
Filas coincidentes después de eliminar customerID: 22


### Validación de registros coincidentes

El dataset original no contiene registros completamente duplicados y cada valor de `customerID` identifica de manera única a un cliente.

Después de eliminar `customerID`, se identificaron 22 filas coincidentes en las variables restantes. Estas filas no se eliminan, ya que corresponden a clientes distintos que presentan las mismas características y no a duplicados reales del dataset original.

Por lo tanto, se conservan los 7043 registros originales.